### PR TITLE
Parser

### LINKED ISSUE #3: Implement parser in scope needed for example
Implement a Parser being able to parse into AST with nodes from #2

### PR DESCRIPTION
**Pull request by jmolinski**
I'm sorry

### CODE REVIEW

File: src/nodes/nodes.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+type CondExpr struct {
+	Test  Node
+	Expr1 Node
+	Expr2 Node
+	NodeCommon
+}
+
+type Or struct {
```
Comment (badochov) [inline src/nodes/nodes.go]: I would remove BinExpr descendant and UnaryExpr descendandt in favour of conrespondingly BinExpr with operator and UnaryExpr with operator should simplify the code further down the road

File: src/parser/parser.go [hunk 2]
```diff
@@ -0,0 +1,1228 @@
...
+		return expr, nil
+	}
+	return &nodes.Compare{
+		Expr:       expr,
+		Ops:        ops,
+		NodeCommon: nodes.NodeCommon{Lineno: lineno},
+	}, nil
+}
+
+func (p *parser) parseMath1() (nodes.Node, error) {
```
Comment (badochov) [inline src/parser/parser.go:423]: id rename to `parseMathAdd`

File: src/parser/parser.go [hunk 3]
```diff
@@ -0,0 +1,1228 @@
...
+		if err != nil {
+			return nil, err
+		}
+		left = makeMathNode(left, right, currentType, lineno)
+		lineno = p.stream.Current().Lineno
+	}
+	return left, nil
+}
+
+func (p *parser) parseMath2() (nodes.Node, error) {
```
Comment (badochov) [inline src/parser/parser.go:443]: id rename to `parseMathMul`

File: src/parser/parser.go [hunk 4]
```diff
@@ -0,0 +1,1228 @@
...
+	lineno := p.stream.Current().Lineno
+	left, err := p.parsePow()
+	if err != nil {
+		return nil, err
+	}
+
+	for p.stream.Current().Type == lexer.TokenMul || p.stream.Current().Type == lexer.TokenDiv || p.stream.Current().Type == lexer.TokenFloordiv || p.stream.Current().Type == lexer.TokenMod {
```
Comment (badochov) [inline src/parser/parser.go]: can be changed to `slices.COntains([]string{alltokens}, p.stream.Current().Type)`

Comment (badochov) [inline src/parser/parser.go]: can be changed to `slices.COntains([]string{alltokens}, p.stream.Current().Type)`

File: src/parser/parser.go [hunk 5]
```diff
@@ -0,0 +1,1228 @@
...
+		Items:      args,
+		Ctx:        "load",
+		NodeCommon: nodes.NodeCommon{Lineno: lineno},
+	}, nil
+}
+
+func (p *parser) parsePrimary() (nodes.Node, error) {
+	token := p.stream.Current()
+	var node nodes.Node
+	if token.Type == lexer.TokenName {
```
Comment (badochov) [inline src/parser/parser.go]: swicth is better match here especially that simetime one branch correspondds to two different tokens

File: src/parser/parser.go [hunk 6]
```diff
@@ -0,0 +1,1228 @@
...
+		if token.Value == "true" || token.Value == "false" || token.Value == "True" || token.Value == "False" {
```
Comment (badochov) [inline src/parser/parser.go]: switch here as well

File: src/parser/parser.go [hunk 7]
```diff
@@ -0,0 +1,1228 @@
...
+				Nodes:      dataBuffer,
+				NodeCommon: nodes.NodeCommon{Lineno: lineno},
+			})
+			dataBuffer = make([]nodes.Node, 0)
+		}
+	}
+
+	for p.stream.Bool() {
+		token := p.stream.Current()
+		if token.Type == lexer.TokenData {
```
Comment (badochov) [inline src/parser/parser.go]: can be changed to switch

File: src/parser/parser.go [hunk 8]
```diff
@@ -0,0 +1,1228 @@
...
+	"with",
+	"autoescape",
+)
+
+var compareOperators = set.FrozenFromElems(
+	"eq", "ne", "lt", "lteq", "gt", "gteq",
+)
+
+func makeMathNode(left, right nodes.Node, op string, lineno int) nodes.Node {
+	if op == "add" {
```
Comment (badochov) [inline src/parser/parser.go]: can be changed to switch

File: src/parser/parser.go [hunk 9]
```diff
@@ -0,0 +1,1228 @@
...
+	//            node = nodes.Neg(self.parse_unary(False), lineno=lineno)
+	//        elif token_type == "add":
+	//            next(self.stream)
+	//            node = nodes.Pos(self.parse_unary(False), lineno=lineno)
+	//        else:
+	//            node = self.parse_primary()
+	//        node = self.parse_postfix(node)
+	//        if with_filter:
+	//            node = self.parse_filter_expr(node)
+	//        return node
```
Comment (badochov) [inline src/parser/parser.go]: this comment should be removed

File: src/parser/parser.go [hunk 10]
```diff
@@ -0,0 +1,1228 @@
...
+
+	lineno := p.stream.Current().Lineno
+	tokenType := p.stream.Current().Type
+	var node nodes.Node
+	var err error
```
Comment (badochov) [inline src/parser/parser.go]: we can use named return to get rid of those variable declarations

File: src/parser/parser.go [hunk 11]
```diff
@@ -0,0 +1,1228 @@
...
+
+func (p *parser) parseCallArgs() ([]nodes.Node, []nodes.Node, *nodes.Node, *nodes.Node, error) {
+	token, err := p.stream.Expect(lexer.TokenLParen)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	var args []nodes.Node
+	var kwargs []nodes.Node
+	var dynArgs *nodes.Node
+	var dynKwargs *nodes.Node
```
Comment (badochov) [inline src/parser/parser.go]: can be converted to named return

File: src/parser/parser.go [hunk 12]
```diff
@@ -0,0 +1,1228 @@
...
+		var dynArgs *nodes.Node
+		var dynKwargs *nodes.Node
+		if p.stream.Current().Type == lexer.TokenLParen {
+			args, kwargs, dynArgs, dynKwargs, err = p.parseCallArgs()
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			args = []nodes.Node{}
+			kwargs = []nodes.Node{}
```
Comment (badochov) [inline src/parser/parser.go]: I see no reason for this initialization

File: src/parser/parser.go [hunk 13]
```diff
@@ -0,0 +1,1228 @@
...
+}
+
+func (p *parser) parseDict() (nodes.Node, error) {
+	// TODO
+	panic("not implemented")
+}
+
+func (p *parser) isTupleEnd(extraEndRules []string) bool {
+	current := p.stream.Current()
+	if current.Type == lexer.TokenVariableEnd || current.Type == lexer.TokenBlockEnd || current.Type == lexer.TokenRParen {
```
Comment (badochov) [inline src/parser/parser.go]: possibly `slice.Contains`

File: src/parser/parser.go [hunk 14]
```diff
@@ -0,0 +1,1228 @@
...
+			messages = append(messages, "You probably made a nesting mistake. Jinja is expecting this tag, but currently looking for "+*currentlyLooking+".")
+		} else {
+			messages = append(messages, "Jinja was looking for the following tags: "+*currentlyLooking+".")
+		}
+	}
+
+	if p.tagStack.Len() > 0 {
+		lastTag := *p.tagStack.Peek()
+		messages = append(messages, fmt.Sprintf("The innermost block that needs to be closed is %q.", lastTag))
+	}
```
Comment (badochov) [inline src/parser/parser.go]:
```go
if lastTag := p.tagStack.Peek(); lastTag != nil {
    		messages = append(messages, fmt.Sprintf("The innermost block that needs to be closed is %q.", *lastTag))

}
```

File: src/utils/stack/stack.go [hunk 15]
```diff
@@ -26,3 +26,11 @@ func (stack *Stack[K]) Pop() *K {
...
 	*stack = append(*stack, el)
 }
+
+func (stack *Stack[K]) AsSlice() []K {
+	return *stack
+}
+
+func (stack *Stack[K]) Len() int {
+	return len(*stack)
+}
```
Comment (badochov) [inline src/utils/stack/stack.go]: I am not a fun of this additions I think method `Iter() []K` would be sufficient

**Review by badochov — CHANGES_REQUESTED** [review]
Magnificient job, though I have some issues with style, and I am not sure if current style of BinExpr and UnaryExpr nodes is good

File: src/parser/parser.go (hunk 2 above)
Comment (jmolinski) [reply src/parser/parser.go:423]: You sure? that's jinja's naming, do we want to change it?

File: src/parser/parser.go (hunk 9 above)
Comment (jmolinski) [reply src/parser/parser.go]: I wonder how it made it here, I must have collapsed it in the IDE...

File: src/utils/stack/stack.go (hunk 15 above)
Comment (jmolinski) [reply src/utils/stack/stack.go]: How would `Iter()` differ from `AsSlice()`?

Comment (badochov) [reply src/utils/stack/stack.go]: implemention wise its the same just something else when it coems to conception

File: src/nodes/nodes.go (hunk 1 above)
Comment (jmolinski) [reply src/nodes/nodes.go]: That's a good idea

File: src/nodes/nodes.go [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
```
Comment (badochov) [inline src/nodes/nodes.go]: Why arent logic nodes also BinExpr or UnaryExpr?

File: src/nodes/nodes.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+	Right Node
+	NodeCommon
+}
+
+type Not struct {
+	Node Node
+	NodeCommon
+}
+
+type Operand struct {
```
Comment (badochov) [inline src/nodes/nodes.go:63]: how does it differ from UnaryOp?

**Review by badochov — COMMENTED** [review]
Look's good im just not sure about hose two nodes types

File: src/nodes/nodes.go (hunk 17 above)
Comment (jmolinski) [reply src/nodes/nodes.go:63]: I guess it doesn't, but jinja differentiates them (looks like Operand is used for chained comparisons?), and operand is not const folded

File: src/nodes/nodes.go (hunk 16 above)
Comment (jmolinski) [reply src/nodes/nodes.go]: I converted them to BinOp and UnaryOp, but a thing to note is that they don't have defined tokens in lexer (I've added comments in nodes.go)

**Review by badochov — APPROVED** [review]
LGTM

**Pull request merged**